src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1580,138 +1625,90 @@ private String getTokenHash(String token) {
+            // shouldn't create the refresh item.
             return null;
         }
         
-        if (userIdentifierType == UserIdentifierType.UniqueId) {
-            Logger.v(TAG + methodName, "UserIdentifier type is unique id, will look for the family item based on unique id.");
-            while (allItems.hasNext()) {
-                final TokenCacheItem tokenCacheItem = allItems.next();
-                final UserInfo tokenUserInfo = tokenCacheItem.getUserInfo();
-                if (tokenUserInfo != null && userId.equalsIgnoreCase(tokenUserInfo.getUserId())) {
-                    if (!StringExtensions.IsNullOrBlank(tokenCacheItem.getFamilyClientId())
-                            && !StringExtensions.IsNullOrBlank(tokenCacheItem.getRefreshToken())) {
-                        Logger.v(TAG + methodName, "Found family item matching the given user id, and the clientId selected for FoCI is: " + tokenCacheItem.getClientId());
-                        return tokenCacheItem;
-                    }
-                }
-            }
-        } else if (userIdentifierType == UserIdentifierType.LoginHint) {
-            Logger.v(TAG + methodName, "UserIdentifier type is loginhint, will look for the family item based on login hint.");
-            while (allItems.hasNext()) {
-                final TokenCacheItem tokenCacheItem = allItems.next();
-                final UserInfo tokenUserInfo = tokenCacheItem.getUserInfo();
-                if (tokenUserInfo != null && userId.equalsIgnoreCase(tokenUserInfo.getDisplayableId())) {
-                    if (!StringExtensions.IsNullOrBlank(tokenCacheItem.getFamilyClientId())
-                            && !StringExtensions.IsNullOrBlank(tokenCacheItem.getRefreshToken())) {
-                        Logger.v(TAG + methodName, "Found family item matching the given user id, and the clientId selected for FoCI is: " + tokenCacheItem.getClientId());
-                        return tokenCacheItem;
-                    }
-                }
-            }
+        final String keyUsed = CacheKey.createCacheKey(authRequest, keyEntryType, user);
+        if (StringExtensions.IsNullOrBlank(keyUsed)) {
+            return null;
         }
 
-        Logger.v(TAG + methodName, "No family items found in the cache.");
-        return null;
+        return createRefreshItem(authRequest, keyUsed, keyEntryType, mTokenCacheStore.getItem(keyUsed));
     }
     
+    /**
+     * @return {@link RefreshItem}.
+     */
+    private RefreshItem createRefreshItem(final AuthenticationRequest authenticationRequest, final String keyUsed, final KeyEntryType keyEntryType, 
+            final TokenCacheItem tokenCacheItem) {
+        if (tokenCacheItem != null && !StringExtensions.IsNullOrBlank(tokenCacheItem.getRefreshToken())) {
+            final String refreshTokenHash = getTokenHash(tokenCacheItem.getRefreshToken());
+
+            Logger.v(TAG, "Refresh token is available and id:" + refreshTokenHash
```

Comment:
It may help for debugging purpose, we can tell if the token is what we actually want. It's not good for us to log the token directly.